### PR TITLE
Unit tests for Enumerable.Where

### PR DESCRIPTION
**Pull request by hpshelton**
Following MarcinJuraszek's examples for Enumerable.Select, I added basically identical tests for Enumerable.Where.

### CODE REVIEW

File: src/System.Linq/src/Var.cs [hunk 1]
```diff
@@ -1,5 +1,4 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
-internal static class var { } // Enforce explicit declaration of variable types.
-
 
+internal static class var { } // Enforce explicit declaration of variable types.
```
Comment (stephentoub) [inline src/System.Linq/src/Var.cs]: Why the change to this file? (FWIW, I personally think we should get rid of this file, as our coding guidelines do allow for the use of var in some cases, and even if they didn't, eventually a Roslyn analyzer would be the preferred way to prevent its usage.  But if we were to get rid of it, we'd want to do it separately from test improvements.)

File: src/System.Linq/tests/WhereTests.cs [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class WhereTests
+    {
+        #region Null arguments
+
+        [Fact]
+        public void Where_Null_ArgumentNullExceptionThrown()
```
Comment (stephentoub) [inline src/System.Linq/tests/WhereTests.cs]: Maybe "SourceIsNull" or "NullSource" instead of just "Null"?

File: src/System.Linq/tests/WhereTests.cs [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+
+            // Ensure subsequent MoveNext calls succeed
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal(2, enumerator.Current);
+        }
+
+        /// <summary>
+        /// Test enumerator - returns int values from 1 to 5 inclusive.
+        /// </summary>
+        private class TestEnumerator : IEnumerable<int>, IEnumerator<int>
```
Comment (stephentoub) [inline src/System.Linq/tests/WhereTests.cs]: Can this be factored out into a project-wide helper?  Looks the same as the one used in the Select tests.

Comment (stephentoub) [inline src/System.Linq/tests/WhereTests.cs]: (Same question for most of the derived enumerators, too.)

File: src/System.Linq/tests/WhereTests.cs [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+
+            using (var enumerator1 = result.GetEnumerator())
+            using (var enumerator2 = result.GetEnumerator())
+            {
+                Assert.Same(result, enumerator1);
+                Assert.NotSame(enumerator1, enumerator2);
+            }
+        }
+    }
+}
```
Comment (stephentoub) [inline src/System.Linq/tests/WhereTests.cs]: Nit: missing newline

File: src/System.Linq/tests/WhereTests.cs [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+            IEnumerable<int> result = source.Where(truePredicate);
+
+            var enumerator = result.GetEnumerator();
+            while (enumerator.MoveNext()) ;
+
+            Assert.Equal(default(int), enumerator.Current);
+        }
+
+        [Fact]
+        public void WhereWhere_Array_ReturnsExpectedValues()
```
Comment (stephentoub) [inline src/System.Linq/tests/WhereTests.cs]: The "WhereWhere" combination is interesting because Where creates a custom Iterator for Where which is then recognized by subsequent Where calls.  Potentially more interesting though is the "WhereSelect" combination, as there are special code paths to handle that common pattern.  Would you consider adding some tests for that combination, too?

There's also one other Where overload, that takes a Func<TSource, int, bool> instead of a Func<TSource, bool>.  Do you have any plans to add tests for that overload, too?

**Comment by stephentoub** [comment]
Thanks for adding these, @hpshelton.  I left a few comments / questions, but overall it looks good to me.

cc: @AlekseyTs

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Linq/tests/WhereTests.cs (hunk 2 above)
Comment (hpshelton) [inline src/System.Linq/tests/WhereTests.cs]: Find-replace mistake. Fixed.

File: src/System.Linq/tests/WhereTests.cs (hunk 3 above)
Comment (hpshelton) [inline src/System.Linq/tests/WhereTests.cs]: Sure, that makes sense once both sets of tests are checked in.

File: src/System.Linq/src/Var.cs (hunk 1 above)
Comment (hpshelton) [inline src/System.Linq/src/Var.cs]: I just added a newline between the copyright and code because I'm pedantic about style. I can revert it.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Linq/tests/WhereTests.cs (hunk 5 above)
Comment (hpshelton) [inline src/System.Linq/tests/WhereTests.cs]: Added tests for WhereSelect, SelectWhere (though these may need to move to the Select tests), and the overload.

**Comment by hpshelton** [comment]
@dotnet-bot test this please

File: src/System.Linq/tests/WhereTests.cs [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+        private static readonly Func<int, bool> s_falsePredicate = (value) => false;
+        private static readonly Func<int, int, bool> s_falsePredicateWithIndex = (value, index) => false;
+
+        private static readonly Func<int, bool> s_nullPredicate = null;
+        private static readonly Func<int, int, bool> s_nullPredicateWithIndex = null;
+        
+        private static readonly Func<int, bool> s_evenPredicate = (value) => value % 2 == 0;
+        private static readonly Func<int, bool> s_oddPredicate = (value) => value % 2 == 1;
+
+        private static readonly Func<int, int, bool> s_complexPredicateWithIndex = (value, index) => { return (value % 2 == 0 && index % 2 == 0); };
```
Comment (stephentoub) [inline src/System.Linq/tests/WhereTests.cs]: IMHO, separating out the lambda like this makes it more difficult to read / understand the test, as it's now relying on functionality that's not local to the test.

**Comment by stephentoub** [comment]
Thanks for adding the additional tests.  Am I correct in my reading, though, that the tests for the Func<TSource, int, bool> overloads will all pass if Where incorrectly always passed 0 as the index?

File: src/System.Linq/tests/WhereTests.cs (hunk 6 above)
Comment (hpshelton) [inline src/System.Linq/tests/WhereTests.cs]: I went back and forth on that. Reverted.

**Comment by hpshelton** [comment]
Yes, you are. I've fixed the complex predicate to `value == index` to validate the index value is correct in addition to testing the predicate is applied correctly.

**Comment by hpshelton** [comment]
Squashed commits. Any final comments?

**Comment by stephentoub** [comment]
LGTM. Thanks!
